File: sdk/samples/08_interop/python/rti_interop.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
rti_interop.py — HDDS subscriber with RTI Connext-compatible QoS

Subscribes on "InteropTest" using QoS.rti_defaults() for wire
compatibility with RTI Connext DDS.

Run:
    python3 rti_interop.py

RTI Connext peer: see peer_commands.md

Expected:
    Received 52 bytes: id=1, msg="Hello from RTI #1"
    ...
"""

import sys
import struct

sys.path.insert(0, "../../../python")
import hdds


def deserialize_string_msg(data: bytes) -> tuple:
    """Deserialize StringMsg {id: u32, message: string} from CDR LE."""
    if len(data) < 8:
        return (0, "")
    msg_id, slen = struct.unpack_from("<II", data, 0)
    if slen == 0 or 8 + slen > len(data):
        return (msg_id, "")
    text = data[8 : 8 + slen - 1].decode("utf-8", errors="replace")
    return (msg_id, text)


def main() -> None:
    hdds.logging.init(hdds.LogLevel.INFO)
    participant = hdds.Participant("RTI_Interop")
    qos = hdds.QoS.rti_defaults()
    reader = participant.create_reader("InteropTest", qos=qos)

    waitset = hdds.WaitSet()
    cond = reader.get_status_condition()
    waitset.attach(cond)

    print("[HDDS] Subscribing on 'InteropTest' (RTI-compatible QoS)...")
    print("[HDDS] Start an RTI Connext publisher on the same topic.\n")

    received = 0
    for _ in range(60):
        if waitset.wait(timeout=1.0):
            while True:
                data = reader.take()
                if data is None:
                    break
                msg_id, text = deserialize_string_msg(data)
                print(f'Received {len(data)} bytes: id={msg_id}, msg="{text}"')
                received += 1

    print(f"\nReceived {received} total messages.")


if __name__ == "__main__":
    main()

File: sdk/samples/08_interop/c/rti_interop.c
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

/*
 * rti_interop.c — HDDS subscriber with RTI Connext-compatible QoS
 *
 * Subscribes on "InteropTest" using hdds_qos_rti_defaults() which
 * configures RTPS wire-format parameters to match RTI Connext defaults.
 * Run an RTI Connext publisher on the same domain/topic to send data.
 *
 * Build:
 *   gcc -o rti_interop rti_interop.c -I../../../c/include -lhdds
 *
 * Run:
 *   ./rti_interop
 *
 * RTI Connext peer: see peer_commands.md
 *
 * Expected:
 *   Received 64 bytes: id=1, msg="Hello from RTI #1"
 */

#include <hdds.h>
#include <stdio.h>
#include <string.h>

/* Deserialize StringMsg {id: u32, message: string} from CDR LE. */
static int deser(const uint8_t *buf, size_t len, uint32_t *id, char *msg, size_t cap) {
    if (len < 8) return -1;
    memcpy(id, buf, 4);
    uint32_t sl; memcpy(&sl, buf + 4, 4);
    if (sl == 0 || 8 + sl > len) return -1;
    size_t n = (sl < cap) ? sl : cap - 1;
    memcpy(msg, buf + 8, n);
    msg[n] = '\0';
    return 0;
}

int main(void) {
    hdds_logging_init(3);

    HddsParticipant *p = hdds_participant_create("RTI_Interop");
    if (!p) { fprintf(stderr, "Participant creation failed\n"); return 1; }

    HddsQoS *qos = hdds_qos_rti_defaults();
    HddsDataReader *reader = hdds_reader_create_with_qos(p, "InteropTest", qos);
    if (!reader) { fprintf(stderr, "Reader creation failed\n"); return 1; }

    HddsWaitSet *ws = hdds_waitset_create();
    hdds_waitset_attach_status_condition(ws, hdds_reader_get_status_condition(reader));

    printf("[HDDS] Subscribing on 'InteropTest' (RTI-compatible QoS)...\n");
    printf("[HDDS] Start an RTI Connext publisher on the same topic.\n\n");

    int received = 0;
    for (int attempt = 0; attempt < 60; attempt++) {
        const void *trig[1]; size_t tc = 0;
        if (hdds_waitset_wait(ws, 1000000000LL, trig, 1, &tc) == HDDS_OK && tc > 0) {
            uint8_t buf[4096]; size_t len = 0;
            while (hdds_reader_take(reader, buf, sizeof(buf), &len) == HDDS_OK) {
                uint32_t id; char msg[256];
                if (deser(buf, len, &id, msg, sizeof(msg)) == 0)
                    printf("Received %zu bytes: id=%u, msg=\"%s\"\n", len, id, msg);
                else
                    printf("Received %zu bytes (unable to decode)\n", len);
                received++;
            }
        }
    }

    printf("\nReceived %d total messages.\n", received);
    hdds_waitset_destroy(ws);
    hdds_reader_destroy(reader);
    hdds_qos_destroy(qos);
    hdds_participant_destroy(p);
    return 0;
}
